source/Test.cpp

```diff
@@ -192,12 +274,81 @@ void Test::Step(Context &context, UI &menuPanels, UI &gamePanels, PlayerInfo &pl
 		return;
 	}
 	
-	const TestStep &stepToRun = steps[context.stepToRun];
+	// All processing was done just before this step started.
+	context.unprocessedInput = false;
+	context.branchesSinceGameStep.clear();
 	
-	// Exit with error on a failing testStep.
-	const string &stepTypeName = STEPTYPE_TO_TEXT.at(stepToRun.stepType);
-	
-	Fail("Test step " + to_string(context.stepToRun) + " (" + stepTypeName + ") failed");
+	bool continueGameLoop = false;
+	while(context.stepToRun < steps.size() && !continueGameLoop)
+	{
+		// Fail if we encounter a watchdog timeout
+		if(context.watchdog == 1)
+			Fail(context, player, "watchdog timeout");
+		else if(context.watchdog > 1)
+			--(context.watchdog);
+		
+		const TestStep &stepToRun = steps[context.stepToRun];
+		switch(stepToRun.stepType)
+		{
+			case TestStep::Type::APPLY:
+				stepToRun.conditions.Apply(player.Conditions());
+				++(context.stepToRun);
+				break;
+			case TestStep::Type::ASSERT:
+				if(!stepToRun.conditions.Test(player.Conditions()))
+					Fail(context, player, "asserted false");
+				++(context.stepToRun);
+				break;
+			case TestStep::Type::BRANCH:
+				// If we encounter a branch entry twice, then resume the gameloop before the second encounter.
+				if(context.branchesSinceGameStep.count(context.stepToRun))
+				{
+					continueGameLoop = true;
+					break;
+				}
+				if(stepToRun.conditions.Test(player.Conditions()))
+					context.stepToRun = jumpTable.find(stepToRun.jumpOnTrueTarget)->second;
+				else if(!stepToRun.jumpOnFalseTarget.empty())
+					context.stepToRun = jumpTable.find(stepToRun.jumpOnFalseTarget)->second;
+				else
+					++(context.stepToRun);
+				context.branchesSinceGameStep.insert(context.stepToRun);
+				break;
+			case TestStep::Type::INJECT:
+				{
+					// Lookup the data and inject it in the game or into the environment.
+					const TestData* testData = (GameData::TestDataSets()).Get(stepToRun.nameOrLabel);
+					if(!testData->Inject())
+						Fail(context, player, "injecting data failed");
+				}
+				++(context.stepToRun);
+				break;
+			case TestStep::Type::INPUT:
+				if(context.unprocessedInput)
+				{
+					continueGameLoop = true;
+					break;
+				}
+				// Give the relevant inputs here.
+				Fail(context, player, "Input not implemented");
+				// Make sure that we run a gameloop before the next input.
+				context.unprocessedInput = true;
```

Comment:
Would it be sensible to set the "run the game loop" flag here, rather than waiting for the next input or other similar condition?

Reply:
Yes, let's process the input immediately after it was given.